### DISCUSSION TITLE
Newcomer to Open-Source & LLM

### DISCUSSION THREAD

**Discussion by onur-goksu** [ggml-org/llama.cpp#9593] [Q&A]
Hello llama.cpp community,

I'm a medior C/C++ engineer however never worked on open-source projects nor LLMs. I lack the domain language and I'll start from scratch. What is the path like which I should follow?

I couldn't find a similar discussion and wanted to ask the community on this topic. 
Thanks in advance.

**Comment by tbogdala** [comment]
Eureka Labs (https://eurekalabs.ai/) is in process of building content that would be useful to you, but it's not there yet. Those finding this discussion from the future might be best served by that link. 

Otherwise, if you're a C/C++ guy, you'll have to accept that in the year 2024 we still have to know Python to talk to the ML people. A lot of stuff exists that is geared towards Python and Pytorch. If you can get past the goofy syntax and spend a little time in a jupyter notebook with pytorch, that'll help when trying to get situated in something far less documented like GGML. I recommend Karpathy's zero-to-hero playlist: https://www.youtube.com/playlist?list=PLAqhIrjkxbuWI23v9cThsA9GvCAUhRvKZ but the fast.ai courses (https://course.fast.ai) are really great too. Jermey Howard's presentation style is different and it may or may not suit you better.

Based on what I've learned over the last year or so, this field is very _hands on_ and benefits from practical exploration of ideas. At any point, if you feel inspired to try something, I'd take a break from the video courses and go for it. That'll be more rewarding anyway.

By the end of Karpathy's zero-to-hero you can write your own tokenizer and neural net architecture in pytorch and a gpt2 sample implementation is in the GGML repository for when you want to cross back over to C/C++. Going through the Let's Build of gpt2 is definitely some of the most programming fun I've had. I did it for Rust and Candle and now I'm in process of doing it for C and GGML.

**Comment by onur-goksu** [comment]
> Eureka Labs (https://eurekalabs.ai/) is in process of building content that would be useful to you, but it's not there yet. Those finding this discussion from the future might be best served by that link. 
> 
> 
> 
> Otherwise, if you're a C/C++ guy, you'll have to accept that in the year 2024 we still have to know Python to talk to the ML people. A lot of stuff exists that is geared towards Python and Pytorch. If you can get past the goofy syntax and spend a little time in a jupyter notebook with pytorch, that'll help when trying to get situated in something far less documented like GGML. I recommend Karpathy's zero-to-hero playlist: https://www.youtube.com/playlist?list=PLAqhIrjkxbuWI23v9cThsA9GvCAUhRvKZ but the fast.ai courses (https://course.fast.ai) are really great too. Jermey Howard's presentation style is different and it may or may not suit you better.
> 
> 
> 
> Based on what I've learned over the last year or so, this field is very _hands on_ and benefits from practical exploration of ideas. At any point, if you feel inspired to try something, I'd take a break from the video courses and go for it. That'll be more rewarding anyway.
> 
> 
> 
> By the end of Karpathy's zero-to-hero you can write your own tokenizer and neural net architecture in pytorch and a gpt2 sample implementation is in the GGML repository for when you want to cross back over to C/C++. Going through the Let's Build of gpt2 is definitely some of the most programming fun I've had. I did it for Rust and Candle and now I'm in process of doing it for C and GGML.

Thanks a lot! I was already looking for some courses to begin with AI. I'll definitely check out these resources.

And also I wouldn't say that I won't use any other programming language. I know Python is a must when it comes to ML. Maybe I should start with Python and other commonly used tools to grasp the domain knowledge first.
